sirepo/package_data/static/js/sirepo.js

```diff
@@ -1556,14 +1544,154 @@ SIREPO.app.service('validationService', function(utilities) {
 
 });
 
+SIREPO.app.factory('srCache', function(appState, $rootScope) {
+
+    // Browser side caching implemented using indexedDB
+    // - Caches sim frame responses
+    // - Allows cache to be cleared by simId or (simId, modelName)
+    // - Keeps MRU date on item access and deletes expired records at startup
+
+    const self = {};
+    const STORE = 'db';
+    const FRAME = 'frame';
+    // 30 days until expired
+    const EXPIRY_TIME = 30 * 24 * 60 * 60 * 1000;
+    let db = null;
+
+    function deleteKeys(keys) {
+        if (keys.length) {
+            const o = objectStore('readwrite');
+            if (o) {
+                for (const k of keys) {
+                    o.delete(k);
+                }
+            }
+        }
+    }
+
+    function invokeCallback(callback, value) {
+        $rootScope.$applyAsync(() => callback(value));
+    }
+
+    function key(prefix, value) {
+        return prefix + ':' + value;
+    }
+
+    function objectStore(mode) {
+        try {
+            if (db) {
+                return db.transaction(STORE, mode).objectStore(STORE);
+            }
+        }
+        catch (e) {
+        }
+        return null;
+    }
 
-SIREPO.app.factory('frameCache', function(appState, panelState, requestSender, authState, $interval, $rootScope, $timeout) {
+    function open() {
+        if (! window.indexedDB || ! SIREPO.authState.uiWebSocket) {
+            return;
+        }
+        const r = window.indexedDB.open('srCache', 1);
+        r.onsuccess = (event) => {
+            db = event.target.result;
+            removeOldRecords();
+        };
+        r.onupgradeneeded = (event) => {
+            db = event.target.result;
+            if (db.objectStoreNames.contains(STORE)) {
+                db.deleteObjectStore(STORE);
+            }
+            const o = db.createObjectStore(STORE);
+            o.createIndex('simId', '_srcache_simId', { unique: false });
+            o.createIndex('mru', '_srcache_mru', { unique: false });
+        };
+    }
+
+    function removeOldRecords() {
+        const o = objectStore('readonly');
+        if (! o) {
+            return;
+        }
+        const expired = [];
+        const d = new Date().getTime();
+        o.index('mru').openKeyCursor().onsuccess = (event) => {
+            const c = event.target.result;
+            if (c) {
+                if ((d - c.key) > EXPIRY_TIME) {
+                    expired.push(c.primaryKey);
+                }
+                c.continue();
+            }
+            else {
+                deleteKeys(expired);
+            }
+        };
+    }
+
+    self.clearFrames = (simId, modelName) => {
```

Comment:
Having frames and generic "object" seems odd to me. This is srFrameCache or something like that.